### ISSUE TITLE
Question: Online ASR example unsuitable for production

### ISSUE THREAD

**Issue by rschluesselbauer** [NVIDIA/NeMo#555]
Hello,
in your tutorial (https://github.com/NVIDIA/NeMo/blob/master/examples/asr/notebooks/2_Online_ASR_Microphone_Demo.ipynb) you write that this type of online decoding is unsuitable for production workflows.
Apart from the lack of language model support, could you please shortly explain the issues one would face?
Thanks

**Comment by okuchaiev** [comment]
You can handle prototypes and, perhaps, even low traffic scenarios with an approach outlined in the notebook. However, when scalability and performance (throughput, latency) are important a deployment solution such as Jarvis (https://developer.nvidia.com/nvidia-jarvis ) should be used.

**Comment by mdda** [comment]
Sorry to hijack this issue - but could you give me some intuition about where `offset=4` is coming from in the initialisation of `FrameASR()` in this tutorial?   It seems like the central window is carefully selected from the middle of the returned data, but then 4 of the output logits are dropped from one end only.  

I'm interested in getting accurate timing for the frames out, and this factor has me puzzled.

**Comment by vsl9** [comment]
Martin, there is the following line in FrameASR's constructor (please notice  "**- 2**"):
```
self.n_timesteps_overlap = int(frame_overlap / timestep_duration) - 2
```
And then we take the central window in `_decode` method:
```
logits[self.n_timesteps_overlap:-self.n_timesteps_overlap]
```
It is slightly wider than required, so we need to drop 4 extra characters. It can be done either way (4 from one end or 2 from both left and right).
With some specific `frame_len` and `frame_overlap`, one might want to tune offset a little bit.

**Comment by tugstugi** [comment]
@vsl9 I want also hijack this issue :) Where do the values for `fixed_mean` and `fixed_std` come from? How to compute them i.e. for smaller nets like QuartzNet5x5?

**Comment by vsl9** [comment]
> Where do the values for fixed_mean and fixed_std come from?

@tugstugi, `fixed_mean` and `fixed_std` were computed as means and standard deviations of mel-spectrogram channels on training dataset.
We normalize each example separately during training and offline inference. Since online inference works on infinite audio stream, we need to use either fixed or adaptive normalization. The former is simpler for the notebook.

> How to compute them i.e. for smaller nets like QuartzNet5x5?

These values don't depend on model's size. The only relevant parameter is a number of mel channels (`features` parameter in YAML): https://github.com/NVIDIA/NeMo/blob/71e40ff32ef1352692ddee256781f58f86b699d8/examples/asr/configs/quartznet15x5.yaml#L21 

If you would like to re-compute these values on your own dataset (or with another `features` value), please use audio features extraction code (as in data layer).

**Comment by mdda** [comment]
Thanks for the answer about `offset`.  However, looking at the sizes of the various frames, etc, it seems that you may be getting an extra mel out of your signal (i.e. 451 instead of the 450=4.5sec/0.01sec expected) - probably due to a fence-posts vs fences thing w.r.t the windowing.  So, the `offset` should probably be 5 here (so that the output number of chars or logits per 0.500sec is exactly 25, rather than 26, for `timestep_duration=0.020sec` ).

**Comment by ArtemisZGL** [comment]
@vsl9 hello, what if I use the Chinese model, do I need to recompute those values?

**Comment by ridasaleem0** [comment]
> @vsl9 hello, what if I use the Chinese model, do I need to recompute those values?

@vsl9 were you able to use it for the Chinese model? I have the same question.
